frame/nomination-pools/src/tests.rs

```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
+			},
+			4 => {
+				// pool_withdraw_unbonded
+				let pool_id = random_existing_pool(&mut rng).unwrap_or_default();
+				let (origin, _) = random_signed_origin(&mut rng);
+				(PoolsCall::<T>::pool_withdraw_unbonded { pool_id, num_slashing_spans: 0 }, origin)
+			},
+			5 => {
+				// withdraw_unbonded
+				let (origin, who) = random_signed_origin(&mut rng);
+				(
+					PoolsCall::<T>::withdraw_unbonded {
+						member_account: who,
+						num_slashing_spans: 0,
+					},
+					origin,
+				)
+			},
+			6 => {
+				// create
+				let (origin, who) = random_signed_origin(&mut rng);
+				let amount = random_ed_multiple(&mut rng);
+				fund_account(&mut rng, &who);
+				let root = who.clone();
+				let state_toggler = who.clone();
+				let nominator = who.clone();
+				(PoolsCall::<T>::create { amount, root, state_toggler, nominator }, origin)
+			},
+			7 => {
+				// nominate
+				let (origin, _) = random_signed_origin(&mut rng);
+				let pool_id = random_existing_pool(&mut rng).unwrap_or_default();
+				let validators = Default::default();
+				(PoolsCall::<T>::nominate { pool_id, validators }, origin)
+			},
+			_ => unreachable!(),
+		}
+	}
+
+	#[derive(Default)]
+	struct RewardAgent {
+		who: AccountId,
+		pool_id: Option<PoolId>,
+		expected_reward: Balance,
+	}
+
+	// TODO: inject some slashes into the game.
+	impl RewardAgent {
+		fn new(who: AccountId) -> Self {
+			Self { who, ..Default::default() }
+		}
+
+		fn join(&mut self) {
+			if self.pool_id.is_some() {
+				return
+			}
+			let pool_id = LastPoolId::<T>::get();
+			let amount = 10 * ExistentialDeposit::get();
+			let origin = Origin::signed(self.who);
+			let _ = Balances::deposit_creating(&self.who, 10 * amount);
+			self.pool_id = Some(pool_id);
+			log::info!(target: "reward-agent", "🤖 reward agent joining in {} with {}", pool_id, amount);
+			assert_ok!(PoolsCall::join::<T> { amount, pool_id }.dispatch_bypass_filter(origin));
+		}
+
+		fn claim_payout(&mut self) {
+			// 10 era later, we claim our payout. We expect our income to be roughly what we
+			// calculated.
+			if !PoolMembers::<T>::contains_key(&self.who) {
+				log!(warn, "reward agent is not in the pool yet, cannot claim");
+				return
+			}
+			let pre = Balances::free_balance(&42);
+			let origin = Origin::signed(42);
+			assert_ok!(PoolsCall::<T>::claim_payout {}.dispatch_bypass_filter(origin));
+			let post = Balances::free_balance(&42);
+
+			let income = post - pre;
+			log::info!(
+				target: "reward-agent", "🤖 CLAIM: actual: {}, expected: {}",
+				income,
+				self.expected_reward,
+			);
+			assert_eq_error_rate!(income, self.expected_reward, 10);
+			self.expected_reward = 0;
+		}
+	}
+
+	#[test]
+	fn fuzz_test() {
+		let mut reward_agent = RewardAgent::new(42);
+		sp_tracing::try_init_simple();
+		// NOTE: use this to get predictable (non)randomness:
+		// use::{rngs::SmallRng, SeedableRng};
+		// let mut rng = SmallRng::from_seed([0u8; 32]);
+		let mut rng = thread_rng();
```

Comment:
Good follow up:

From experience I can tell you that it is a good idea to create a random seed, print it, and add the possibility to overide it via env variable.  
Then you always have a different set of tests, but if it fails, you can reproduce by using the printed seed.

Reply:
excellent idea, I actually have this in my code as a commented code that: If you want reproducible output, do this and that, but using a seed and env var makes a whole lot more sense.